File: pandora/cost_volume_confidence/risk.py
#!/usr/bin/env python
# coding: utf8
#
# Copyright (c) 2020 Centre National d'Etudes Spatiales (CNES).
#
# This file is part of PANDORA
#
#     https://github.com/CNES/Pandora_pandora
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
This module contains functions for estimating the risk.
"""

import warnings
from typing import Dict, Tuple, Union

import numpy as np
from json_checker import Checker, And
from numba import njit, prange
import xarray as xr


from . import cost_volume_confidence


@cost_volume_confidence.AbstractCostVolumeConfidence.register_subclass("risk")
class Risk(cost_volume_confidence.AbstractCostVolumeConfidence):
    """
    Allows to estimate a risk confidence from the cost volume
    """

    # Default configuration, do not change this value
    _ETA_MIN = 0.0
    _ETA_MAX = 0.7
    _ETA_STEP = 0.01
    # Percentile value to normalize ambiguity
    _PERCENTILE = 1.0

    def __init__(self, **cfg: str) -> None:
        """
        :param cfg: optional configuration, {'confidence_method': 'risk', 'eta_min': float, 'eta_max': float,
        'eta_step': float}
        :type cfg: dict
        :return: None
        """
        self.cfg = self.check_conf(**cfg)
        self._eta_min = self._ETA_MIN
        self._percentile = self._PERCENTILE
        self._eta_step = float(self.cfg["eta_step"])
        self._eta_max = float(self.cfg["eta_max"])

    def check_conf(self, **cfg: Union[str, float]) -> Dict[str, Union[str, float]]:
        """
        Add default values to the dictionary if there are missing elements and check if the dictionary is correct

        :param cfg: ambiguity configuration
        :type cfg: dict
        :return cfg: ambiguity configuration updated
        :rtype: dict
        """
        if "eta_max" not in cfg:
            cfg["eta_max"] = self._ETA_MAX
        if "eta_step" not in cfg:
            cfg["eta_step"] = self._ETA_STEP

        schema = {
            "confidence_method": And(str, lambda input: "risk"),
            "eta_max": And(float, lambda input: 0 < input < 1),
            "eta_step": And(float, lambda input: 0 < input < 1),
        }

        checker = Checker(schema)
        checker.validate(cfg)
        return cfg

    def desc(self) -> None:
        """
        Describes the confidence method
        :return: None
        """
        print("Risk method")

    def confidence_prediction(
        self,
        disp: xr.Dataset,
        img_left: xr.Dataset = None,
        img_right: xr.Dataset = None,
        cv: xr.Dataset = None,
    ) -> Tuple[xr.Dataset, xr.Dataset]:
        """
        Computes a risk confidence measure that evaluates the matching cost function at each point

        :param disp: the disparity map dataset
        :type disp: xarray.Dataset
        :param img_left: left Dataset image
        :tye img_left: xarray.Dataset
        :param img_right: right Dataset image
        :type img_right: xarray.Dataset
        :param cv: cost volume dataset
        :type cv: xarray.Dataset
        :return: the disparity map and the cost volume with new indicators 'risk_max_confidence'
                and 'risk_min_confidence' in the
        :rtype: Tuple(xarray.Dataset, xarray.Dataset)
        """
        # This silences numba's TBB threading layer warning
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore")
            # Obtain sampled_ambiguity, necessary for risk_min computation
            ambiguity_ = cost_volume_confidence.AbstractCostVolumeConfidence(  # type: ignore
                **{"confidence_method": "ambiguity"}  # type: ignore
            )
            _, sampled_ambiguity = ambiguity_.compute_ambiguity_and_sampled_ambiguity(  # type: ignore
                cv["cost_volume"].data, self._eta_min, self._eta_max, self._eta_step
            )
            # Computes risk using numba in parallel for memory and computation time optimization
            risk_max, risk_min = self.compute_risk(
                cv["cost_volume"].data, sampled_ambiguity, self._eta_min, self._eta_max, self._eta_step
            )

        disp, cv = self.allocate_confidence_map("risk_max_confidence", risk_max, disp, cv)
        disp, cv = self.allocate_confidence_map("risk_min_confidence", risk_min, disp, cv)

        return disp, cv

    @staticmethod
    @njit("Tuple((f4[:, :],f4[:, :]))(f4[:, :, :], f4[:, :, :], f4, f4, f4)", parallel=True, cache=True)
    def compute_risk(
        cv: np.ndarray, sampled_ambiguity: np.ndarray, _eta_min: float, _eta_max: float, _eta_step: float
    ) -> Tuple[np.ndarray, np.ndarray]:
        """
        Computes minimum and maximum risk.

        :param cv: cost volume
        :type cv: 3D np.array (row, col, disp)
        :param sampled_ambiguity: sampled cost volume ambiguity
        :type sampled_ambiguity: 3D np.array (row, col, eta)
        :param _eta_min: minimal eta
        :type _eta_min: float
        :param _eta_max: maximal eta
        :type _eta_max: float
        :param _eta_step: eta step
        :type _eta_step: float
        :return: the minimum and maximum risk
        :rtype: Tuple(2D np.array (row, col) dtype = float32, 2D np.array (row, col) dtype = float32)
        """
        #  Miniumum and maximum of all costs, useful to normalize the cost volume
        min_cost = np.nanmin(cv)
        max_cost = np.nanmax(cv)

        n_row, n_col, nb_disps = cv.shape

        etas = np.arange(_eta_min, _eta_max, _eta_step)

        # Numba does not support the np.tile operation
        two_dim_etas = np.repeat(etas, nb_disps).reshape((-1, nb_disps)).T.flatten()

        # Initialize min and max risk integral
        risk_max = np.zeros((n_row, n_col), dtype=np.float32)
        risk_min = np.zeros((n_row, n_col), dtype=np.float32)

        for row in prange(n_row):  # pylint: disable=not-an-iterable
            for col in prange(n_col):  # pylint: disable=not-an-iterable
                # Normalized minimum cost for one point
                normalized_min_cost = (np.nanmin(cv[row, col, :]) - min_cost) / (max_cost - min_cost)

                # If all costs are at nan, set the risk at nan for this point
                if np.isnan(normalized_min_cost):
                    risk_max[row, col] = np.nan
                    risk_min[row, col] = np.nan
                else:
                    normalized_min_cost = np.repeat(normalized_min_cost, nb_disps * etas.shape[0])
                    # Normalized cost volume for one point
                    normalized_cv = (cv[row, col, :] - min_cost) / (max_cost - min_cost)
                    #  Mask nan to -inf to later discard values out of [min; min + eta]
                    normalized_cv[np.isnan(normalized_cv)] = -np.inf
                    normalized_cv = np.repeat(normalized_cv, etas.shape[0])
                    # Initialize all disparties
                    disp_cv = np.arange(nb_disps) * 1.0
                    disp_cv = np.repeat(disp_cv, etas.shape[0])
                    # Remove disparities for every similarity value outside of [min;min+eta[
                    disp_cv[normalized_cv > (normalized_min_cost + two_dim_etas)] = np.nan
                    # Reshape to distinguish each sample's disparity range
                    disp_cv = disp_cv.reshape((nb_disps, etas.shape[0]))
                    # Initialize minimum and maximum disparities
                    min_disp = np.zeros(etas.shape[0])
                    max_disp = np.zeros(etas.shape[0])
                    # Obtain min and max disparities for each sample
                    for i in range(etas.shape[0]):
                        min_disp[i] = np.nanmin(disp_cv[:, i])
                        max_disp[i] = np.nanmax(disp_cv[:, i])
                    # fill mean risks
                    risk_max[row, col] = np.nanmean(max_disp - min_disp)
                    # fill mean min risk. risk min is defined as mean( (1+risk(p,k)) - amb(p,k) )
                    risk_min[row, col] = np.nanmean((1 + (max_disp - min_disp)) - sampled_ambiguity[row, col, :])

        return risk_max, risk_min

    @staticmethod
    @njit(
        "Tuple((f4[:, :],f4[:, :],f4[:, :, :],f4[:, :, :]))(f4[:, :, :], f4[:, :, :], f4, f4, f4)",
        parallel=True,
        cache=True,
    )
    def compute_risk_and_sampled_risk(
        cv: np.ndarray, sampled_ambiguity: np.ndarray, _eta_min: float, _eta_max: float, _eta_step: float
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """
        Computes minimum and maximum risk and sampled_risk.

        :param cv: cost volume
        :type cv: 3D np.array (row, col, disp)
        :param sampled_ambiguity: sampled cost volume ambiguity
        :type sampled_ambiguity: 3D np.array (row, col, eta)
        :param _eta_min: minimal eta
        :type _eta_min: float
        :param _eta_max: maximal eta
        :type _eta_max: float
        :param _eta_step: eta step
        :type _eta_step: float
        :return: the risk
        :rtype: Tuple(2D np.array (row, col) dtype = float32, 2D np.array (row, col) dtype = float32,
                     3D np.array (row, col) dtype = float32, 3D np.array (row, col) dtype = float32)
        """
        #  Miniumum and maximum of all costs, useful to normalize the cost volume
        min_cost = np.nanmin(cv)
        max_cost = np.nanmax(cv)

        n_row, n_col, nb_disps = cv.shape

        etas = np.arange(_eta_min, _eta_max, _eta_step)

        # Numba does not support the np.tile operation
        two_dim_etas = np.repeat(etas, nb_disps).reshape((-1, nb_disps)).T.flatten()

        # Initialize min and max risk integral
        risk_max = np.zeros((n_row, n_col), dtype=np.float32)
        risk_min = np.zeros((n_row, n_col), dtype=np.float32)
        # Initialize min and max sampled risks
        sampled_risk_min = np.zeros((n_row, n_col, etas.shape[0]), dtype=np.float32)
        sampled_risk_max = np.zeros((n_row, n_col, etas.shape[0]), dtype=np.float32)

        for row in prange(n_row):  # pylint: disable=not-an-iterable
            for col in prange(n_col):  # pylint: disable=not-an-iterable
                # Normalized minimum cost for one point
                normalized_min_cost = (np.nanmin(cv[row, col, :]) - min_cost) / (max_cost - min_cost)

                # If all costs are at nan, set the risk at nan for this point
                if np.isnan(normalized_min_cost):
                    sampled_risk_min[row, col, :] = np.nan
                    sampled_risk_max[row, col, :] = np.nan
                    risk_max[row, col] = np.nan
                    risk_min[row, col] = np.nan
                else:
                    normalized_min_cost = np.repeat(normalized_min_cost, nb_disps * etas.shape[0])
                    # Normalized cost volume for one point
                    normalized_cv = (cv[row, col, :] - min_cost) / (max_cost - min_cost)
                    #  Mask nan to -inf to later discard values out of [min; min + eta]
                    normalized_cv[np.isnan(normalized_cv)] = -np.inf
                    normalized_cv = np.repeat(normalized_cv, etas.shape[0])
                    # Initialize all disparties
                    disp_cv = np.arange(nb_disps) * 1.0
                    disp_cv = np.repeat(disp_cv, etas.shape[0])
                    # Remove disparities for every similarity value outside of [min;min+eta[
                    disp_cv[normalized_cv > (normalized_min_cost + two_dim_etas)] = np.nan
                    # Reshape to distinguish each sample's disparity range
                    disp_cv = disp_cv.reshape((nb_disps, etas.shape[0]))
                    # Initialize minimum and maximum disparities
                    min_disp = np.zeros(etas.shape[0])
                    max_disp = np.zeros(etas.shape[0])
                    # Obtain min and max disparities for each sample
                    for i in range(etas.shape[0]):
                        min_disp[i] = np.nanmin(disp_cv[:, i])
                        max_disp[i] = np.nanmax(disp_cv[:, i])
                    # fill sampled max risk
                    sampled_risk_max[row, col, :] += max_disp - min_disp
                    # fill sampled min risk. risk min is defined as ( (1+risk(p,k)) - amb(p,k) )
                    sampled_risk_min[row, col, :] += (1 + (max_disp - min_disp)) - sampled_ambiguity[row, col, :]
                    # fill mean risks
                    risk_max[row, col] = np.nanmean(max_disp - min_disp)
                    # fill mean min risk. risk min is defined as mean( (1+risk(p,k)) - amb(p,k) )
                    risk_min[row, col] = np.nanmean((1 + (max_disp - min_disp)) - sampled_ambiguity[row, col, :])

        return risk_max, risk_min, sampled_risk_max, sampled_risk_min

File: pandora/cost_volume_confidence/cpp/src/risk.cpp
/* Copyright (c) 2025 Centre National d'Etudes Spatiales (CNES).
 *
 * This file is part of PANDORA
 *
 *     https://github.com/CNES/Pandora
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "risk.hpp"
#include "cost_volume_confidence_tools.hpp"
#include <algorithm>
#include <numeric>
#include <cmath>
#include <tuple>

namespace py = pybind11;

py::list compute_risk_and_sampled_risk(
    py::array_t<float> cv,
    py::array_t<float> sampled_ambiguity,
    py::array_t<double> etas,
    int nbr_etas,
    py::array_t<int> grids,
    py::array_t<float> disparity_range,
    bool sample_risk
) {

    py::list to_return = py::list();

    auto r_cv = cv.unchecked<3>();
    auto r_samp_amb = sampled_ambiguity.unchecked<3>();
    auto r_etas = etas.unchecked<1>();
    auto r_grids = grids.unchecked<3>();

    size_t n_row = cv.shape(0);
    size_t n_col = cv.shape(1);
    size_t n_disp = cv.shape(2);

    // Initialize min and max risk integral
    py::array_t<float> risk_min = py::array_t<float>({n_row, n_col});
    py::array_t<float> risk_max = py::array_t<float>({n_row, n_col});
    auto rw_risk_min = risk_min.mutable_unchecked<2>();
    auto rw_risk_max = risk_max.mutable_unchecked<2>();

    py::array_t<float> risk_disp_inf = py::array_t<float>({n_row, n_col});
    py::array_t<float> risk_disp_sup = py::array_t<float>({n_row, n_col});
    auto rw_risk_disp_inf = risk_disp_inf.mutable_unchecked<2>();
    auto rw_risk_disp_sup = risk_disp_sup.mutable_unchecked<2>();

    py::array_t<float> samp_risk_min;
    py::array_t<float> samp_risk_max;
    std::unique_ptr<py::detail::unchecked_mutable_reference<float, 3>> rw_samp_risk_min;
    std::unique_ptr<py::detail::unchecked_mutable_reference<float, 3>> rw_samp_risk_max;

    if (sample_risk) {
        // Initialize min and max sampled risks
        samp_risk_min = py::array_t<float>({n_row, n_col, static_cast<size_t>(nbr_etas)});
        samp_risk_max = py::array_t<float>({n_row, n_col, static_cast<size_t>(nbr_etas)});
        rw_samp_risk_min = std::make_unique<py::detail::unchecked_mutable_reference<float, 3>>(
            samp_risk_min.mutable_unchecked<3>()
        );
        rw_samp_risk_max = std::make_unique<py::detail::unchecked_mutable_reference<float, 3>>(
            samp_risk_max.mutable_unchecked<3>()
        );
    }

    auto [min_cost, max_cost, rw_min_img, _] = min_max_cost(r_cv, n_row, n_col, n_disp);

    float diff_cost = max_cost - min_cost;

    float* normalized_pix_costs = new float[n_disp];
    for (size_t row = 0; row < n_row; ++row) {
        for (size_t col = 0; col < n_col; ++col) {

            // Normalized extremum cost for one point
            float normalized_extremum = (rw_min_img(row, col) - min_cost) / diff_cost;

            // If all costs are at nan, set the risk at nan for this point
            if (std::isnan(normalized_extremum)) {
                rw_risk_min(row, col) = std::numeric_limits<float>::quiet_NaN();
                rw_risk_max(row, col) = std::numeric_limits<float>::quiet_NaN();

                rw_risk_disp_inf(row, col) = std::numeric_limits<float>::quiet_NaN();
                rw_risk_disp_sup(row, col) = std::numeric_limits<float>::quiet_NaN();
                if (!sample_risk)
                    continue;

                for (size_t eta = 0; eta < nbr_etas; ++eta) {
                    rw_samp_risk_min->operator()(
                        row, col, eta
                    ) = std::numeric_limits<float>::quiet_NaN();
                    rw_samp_risk_max->operator()(
                        row, col, eta
                    ) = std::numeric_limits<float>::quiet_NaN();
                }
                continue;
            }

            size_t idx_disp_min = searchsorted(disparity_range, r_grids(0, row, col));
            size_t idx_disp_max = searchsorted(disparity_range, r_grids(1, row, col)) + 1;

            // fill normalized cv for this pixel (mask with +-inf when encountering NaNs)
            for (size_t disp = 0; disp < n_disp; ++disp) {
                float cv_val = r_cv(row, col, disp);
                if (std::isnan(cv_val)) {
                    if (disp >= idx_disp_min && disp < idx_disp_max) {
                        normalized_pix_costs[disp] = -std::numeric_limits<float>::infinity();
                    }
                    else {
                        normalized_pix_costs[disp] = std::numeric_limits<float>::infinity();
                    }
                    continue;
                }
                normalized_pix_costs[disp] = (cv_val - min_cost) / diff_cost;
            }

            float sum_for_min = 0;
            float sum_for_max = 0;
            float sum_for_disp_inf = 0;
            float sum_for_disp_sup = 0;
            for (size_t eta = 0; eta < nbr_etas; ++eta) {
                // Obtain min and max disparities for each sample
                float min_disp_idx = std::numeric_limits<float>::infinity();
                float max_disp_idx = -std::numeric_limits<float>::infinity();
                for (size_t disp = 0; disp < n_disp; ++disp) {
                    if (normalized_pix_costs[disp] > (normalized_extremum + r_etas(eta)))
                        continue;
                    // case normalized_pix_costs[disp] < normalized_extremum is not computed:
                    // every case is > normalized_extremum, only - inf is not dealt with
                    // -inf is for nan disparities in disparity range, that should be considered to increase risk
                    // this is equivalent to put the > normalized_extremum condition, with nan put to normalized_extremum
                    // but gain condition check

                    min_disp_idx = std::min(min_disp_idx, static_cast<float>(disp));
                    max_disp_idx = std::max(max_disp_idx, static_cast<float>(disp));
                }

                int min_index = static_cast<int>(min_disp_idx);
                int max_index = static_cast<int>(max_disp_idx);

                float min_disp = std::numeric_limits<float>::infinity();
                float max_disp = -std::numeric_limits<float>::infinity();
                py::buffer_info buf_info = disparity_range.request();
                float* data_ptr = static_cast<float*>(buf_info.ptr);

                min_disp = data_ptr[min_index];
                max_disp = data_ptr[max_index];
                // add sampled max risk to sum
                float eta_max_disp = max_disp_idx - min_disp_idx;
                // add sampled min risk to sum. risk min is defined as ( (1+risk(p,k)) - amb(p,k) )
                float eta_min_disp = 1 + eta_max_disp - r_samp_amb(row, col, eta);

                // add sampled min and max disp to sum
                sum_for_disp_sup += max_disp;
                sum_for_disp_inf += min_disp;

                sum_for_min += eta_min_disp;
                sum_for_max += eta_max_disp;
                if (sample_risk) {
                    // fill sampled min and max risk
                    rw_samp_risk_min->operator()(row, col, eta) = eta_min_disp;
                    rw_samp_risk_max->operator()(row, col, eta) = eta_max_disp;
                }
            }

            // fill min/max risk for this pixel
            rw_risk_min(row, col) = sum_for_min / nbr_etas;
            rw_risk_max(row, col) = sum_for_max / nbr_etas;

            rw_risk_disp_sup(row, col) = sum_for_disp_sup / nbr_etas;
            rw_risk_disp_inf(row, col) = sum_for_disp_inf /nbr_etas;

        }
    }

    delete[] normalized_pix_costs;

    to_return.append( risk_max );
    to_return.append( risk_min );
    to_return.append( risk_disp_sup );
    to_return.append( risk_disp_inf );
    if (sample_risk) {
        to_return.append( samp_risk_max );
        to_return.append( samp_risk_min );
    }
    return to_return;
}